TheThingsNetwork/lorawan-stack
# Add support for user bookmarks

#### Summary



References https://github.com/TheThingsNetwork/lorawan-stack/issues/6483

This PR introduces the concept of user bookmarks, allowing the user to setup bookmarks to desired entities.

#### Changes



- Add UserBookmarkRegistry 
- Add CLI commands related to the UserBookmarkRegistry
- Update entities registry to call the operation that interacts with the entity's bookmarks.
  - If deleted, soft-deletes the bookmarks which point to the entity.
  - If restored, restores the bookmarks which point to the entity.
  - If purged, purges the bookmarks which point to the entity.
- Add migration to UserBookmark table.

#### Testing



Unit tests, manual testing.

<details>
<summary>Test Steps</summary>

Do the following for each of the entities [ application, client, end-device, gateway, organization ]:

1. Create a bookmark of an user and the entity.`ttn-lw-cli u bm create <usr_id> --entity-ids.application-id <app_id>`
2. Check if the user's bookmark is there. `ttn-lw-cli usr bookmarks list <usr_id>`
3. Delete the bookmark `ttn-lw-cli u bm delete <user_id> --entity-ids.application-id <app_id>`
4. Check that the user's bookmark is **not** there. Both commands below should not yield results.
    - `ttn-lw-cli usr bookmarks list <usr_id> --deleted`
    - `ttn-lw-cli usr bookmarks list <usr_id>`

Do the following with application and one user:

1. Create a bookmark the user to the entity.`ttn-lw-cli u bm create <usr_id> --entity-ids.application-id <app_id>`
2. Delete the entity. `ttn-lw-cli app delete <app_id>`
3. Search for deleted bookmarks of an user `ttn-lw-cli u bm list <usr_id> --deleted`. Have to show up the bookmark to the deleted entity.
4. Restore the entity. `ttn-lw-cli app restore <app_id>`
5. Search for bookmarks of an user `ttn-lw-cli u bm list <usr_id>`. Have to show up the bookmark to the entity.
6. Purge the entity. `ttn-lw-cli app purge <app_id>`
7. Search for deleted bookmarks of an user `ttn-lw-cli u bm list <usr_id> --deleted`. No bookmark should show up.

</details>

##### Regressions



Addition of a new feature, does not modify old behaviors but introduces new ones. So no regressions.

#### Notes for Reviewers



NOTE: when an `user` is deleted its bookmarks are not deleted, just when its purged. 

The reason for this is to avoid the following scenario:
- Bookmark between user and entity.
- Delete entity
- Delete user
- Restore user.

Under the current implementation of `RestoreUserBookmarks` it would restore the bookmark even with the entity still being in a deleted state. This could mitigated by going through all the bookmarks restored and checking if the entity referenced in each bookmark is not deleted, the alternative that I took on this PR is to not delete bookmarks that belong to an user.

The case of `user-X -> user-Y` when the user-Y is deleted the bookmark is soft-deleted.

#### Checklist
